### PR TITLE
Windows: graph cleanup '-removing'

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard (VM) <jhoward@ntdev.microsoft.com>

@darrenstahlmsft PTAL. Adds code to the graphdriver cleanup to delete any left-over `-removing` directories. These might remain on disk if a previous daemon instance was killed without a chance to cleanup.

### CODE REVIEW

**Comment by darrenstahlmsft** [comment]
Not a maintainer, but LGTM 👍

File: daemon/graphdriver/windows/windows.go [hunk 1]
```diff
@@ -412,7 +412,24 @@ func (d *Driver) Put(id string) error {
...
+	items, err := ioutil.ReadDir(d.info.HomeDir)
+	if err != nil {
+		return nil
+	}
+	for _, item := range items {
+		if item.IsDir() && strings.HasSuffix(item.Name(), "-removing") {
+			if err := hcsshim.DestroyLayer(d.info, item.Name()); err != nil {
+				logrus.Warnf("Failed to cleanup %s: %s", item.Name(), err)
+			} else {
+				logrus.Infof("Cleaned up %s", item.Name())
```
Comment (thaJeztah) [inline daemon/graphdriver/windows/windows.go]: Wondering if this is "info" or "debug" information. Looking for other cleanup actions, those seem to use `Debug`;
https://github.com/docker/docker/blob/7c62f58b27f98140ff3d39de166bf2e107e84a65/plugin/blobstore.go#L69
https://github.com/docker/docker/blob/f67d4b897adc9048269d1b91435b6dcbf69e36d1/libcontainerd/container_windows.go#L61

Open to input what would be best 👍

File: daemon/graphdriver/windows/windows.go [hunk 2]
```diff
@@ -412,7 +412,24 @@ func (d *Driver) Put(id string) error {
 }
 
 // Cleanup ensures the information the driver stores is properly removed.
+// We use this opportunity to cleanup any -removing folders which may be
+// still left if the daemon was killed while it was removing a layer.
 func (d *Driver) Cleanup() error {
+	// Note we don't return an error on cleanup failure, but do log a warning.
```
Comment (thaJeztah) [inline daemon/graphdriver/windows/windows.go]: This comment is a bit odd, because the first error below actually doesn't log anything 😅

**Review by thaJeztah — COMMENTED** [review]
left two nits/questions, but LGTM otherwise

File: daemon/graphdriver/windows/windows.go [hunk 3]
```diff
@@ -412,7 +412,24 @@ func (d *Driver) Put(id string) error {
...
+	items, err := ioutil.ReadDir(d.info.HomeDir)
+	if err != nil {
+		return nil
```
Comment (vdemeester) [inline daemon/graphdriver/windows/windows.go]: If we want to return `nil` here, let's at least log the error.

File: daemon/graphdriver/windows/windows.go [hunk 4]
```diff
@@ -412,7 +412,24 @@ func (d *Driver) Put(id string) error {
...
+	}
+	for _, item := range items {
+		if item.IsDir() && strings.HasSuffix(item.Name(), "-removing") {
```
Comment (vdemeester) [inline daemon/graphdriver/windows/windows.go]:
```go
if !item.IsDir() || !strings.HasSuffix(item.Name(), "-removing") {
    continue
}
// […]
```

**Review by vdemeester — CHANGES_REQUESTED** [review]

File: daemon/graphdriver/windows/windows.go (hunk 2 above)
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: That's fair - moved it lower down and clarified.

File: daemon/graphdriver/windows/windows.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: I think these logs are correct here. Will make sense with the updated comment. It's not debugging information.

File: daemon/graphdriver/windows/windows.go (hunk 3 above)
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: Updated. This one is highly unlikely to fail, so returning the error

File: daemon/graphdriver/windows/windows.go (hunk 4 above)
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: Why? IMHO the current way is more understandable.... 😇

Comment (vdemeester) [inline daemon/graphdriver/windows/windows.go]: I have OCD on complexity and *nested if*, that's all 😅.

File: daemon/graphdriver/windows/windows.go (hunk 1 above)
Comment (thaJeztah) [inline daemon/graphdriver/windows/windows.go]: Thanks for looking; I'm ok with keeping it as info then 👍

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐻

**Pull request merged**